src/poetry/repositories/pool.py

```diff
@@ -1,59 +1,77 @@
 from __future__ import annotations
 
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import IntEnum
+from enum import auto
 from typing import TYPE_CHECKING
 
+from poetry.repositories.abstract_repository import AbstractRepository
 from poetry.repositories.exceptions import PackageNotFound
-from poetry.repositories.repository import Repository
 
 
 if TYPE_CHECKING:
     from poetry.core.constraints.version import Version
     from poetry.core.packages.dependency import Dependency
     from poetry.core.packages.package import Package
 
+    from poetry.repositories.repository import Repository
 
-class Pool(Repository):
+
+class RepositoryPriority(IntEnum):
+    # The order of the members below dictates the actual priority. The first member has
+    # top priority.
+    DEFAULT = auto()
+    PRIMARY = auto()
+    SECONDARY = auto()
+
+
+@dataclass
+class PrioritizedRepository:
+    repository: Repository
+    priority: RepositoryPriority
+
+
+class Pool(AbstractRepository):
     def __init__(
         self,
         repositories: list[Repository] | None = None,
         ignore_repository_names: bool = False,
     ) -> None:
-        super().__init__("poetry-pool")
-
-        if repositories is None:
-            repositories = []
-
-        self._lookup: dict[str, int] = {}
-        self._repositories: list[Repository] = []
-        self._default = False
-        self._has_primary_repositories = False
-        self._secondary_start_idx: int | None = None
+        super().__init__("poetry-repository-pool")
+        self._repositories: OrderedDict[str, PrioritizedRepository] = OrderedDict()
+        self._ignore_repository_names = ignore_repository_names
 
+        repositories = repositories or []
         for repository in repositories:
             self.add_repository(repository)
 
-        self._ignore_repository_names = ignore_repository_names
-
     @property
     def repositories(self) -> list[Repository]:
-        return self._repositories
+        unsorted_repositories = list(self._repositories.values())
+        sorted_repositories = sorted(
+            unsorted_repositories, key=lambda prio_repo: prio_repo.priority
+        )
+        return [prio_repo.repository for prio_repo in sorted_repositories]
 
     def has_default(self) -> bool:
-        return self._default
+        return self._contains_priority(RepositoryPriority.DEFAULT)
 
     def has_primary_repositories(self) -> bool:
-        return self._has_primary_repositories
+        return self._contains_priority(RepositoryPriority.PRIMARY)
+
+    def _contains_priority(self, priority: RepositoryPriority) -> bool:
+        return any(
+            prio_repo.priority == priority for prio_repo in self._repositories.values()
+        )
 
     def has_repository(self, name: str) -> bool:
-        return name.lower() in self._lookup
+        return name.lower() in self._repositories
```

Comment:
We could do this in another PR, but in the long term I'd like to add a `RepositoryName` type that represents a normalized name, and centralize all normalization code there.

Reply:
See [278c432](https://github.com/python-poetry/poetry/pull/6669/commits/278c4325e3bb6a4a11e220fd18aab83217583b44) for an attempt. I'm not too happy with it, but let me know what you think.

Reply:
Ah again this breaks plugin...